src/node/mini_miner.cpp

```diff
@@ -0,0 +1,318 @@
+            // UXTO is created by transaction in mempool, add to map.
+            // Note: This will either create a missing entry or add the outpoint to an existing entry
+            m_requested_outpoints_by_txid[outpoint.hash].push_back(outpoint);
+        } else {
+            // This UTXO is either confirmed or not yet submitted to mempool.
+            // If it's confirmed, no bump fee is required.
+            // If it's not yet submitted, we have no information, so return 0.
+            m_bump_fees.emplace(outpoint, 0);
+        }
+    }
+
+    // No unconfirmed UTXOs, so nothing mempool-related needs to be calculated.
+    if (m_requested_outpoints_by_txid.empty()) return;
+
+    // Calculate the cluster and construct the entry map.
+    std::vector<uint256> txids_needed;
+    for (const auto& [txid, _]: m_requested_outpoints_by_txid) {
+        txids_needed.push_back(txid);
+    }
+    const auto cluster = mempool.GatherClusters(txids_needed);
+    if (cluster.empty()) {
+        // An empty cluster means that at least one of the transactions is missing from the mempool
+        // (should not be possible given processing above) or DoS limit was hit.
+        m_ready_to_calculate = false;
+        return;
+    }
+
+    // Add every entry to m_entries_by_txid and m_entries, except the ones that will be replaced.
+    for (const auto& txiter : cluster) {
+        if (!m_to_be_replaced.count(txiter->GetTx().GetHash())) {
+            auto [mapiter, success] = m_entries_by_txid.emplace(txiter->GetTx().GetHash(), MiniMinerMempoolEntry(txiter));
+            m_entries.push_back(mapiter);
+        } else {
+            auto outpoints_it = m_requested_outpoints_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != m_requested_outpoints_by_txid.end()) {
+                // This UTXO is the output of a to-be-replaced transaction. Bump fee is 0; spending
+                // this UTXO is impossible as it will no longer exist after the replacement.
+                for (const auto& outpoint : outpoints_it->second) {
+                    m_bump_fees.emplace(outpoint, 0);
+                }
+                m_requested_outpoints_by_txid.erase(outpoints_it);
+            }
+        }
+    }
+
+    // Build the m_descendant_set_by_txid cache.
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MiniMinerMempoolEntry
+        // will not exist without its ancestor MiniMinerMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        const bool remove{m_to_be_replaced.count(txid) > 0};
+        CTxMemPool::setEntries descendants;
+        mempool.CalculateDescendants(txiter, descendants);
+        Assume(descendants.count(txiter) > 0);
+        for (const auto& desc_txiter : descendants) {
```

Comment:
This has been puzzling me for a while.

It is not clear to me why need to calculate the descendants once more if they are all in the cluster.
Couldn't return from `GatherCluster` a map of tx id and descendants instead?
Would need to first go up, to the outpoints first parents, and then start going down (basically the DAG in map format).

Another point is why continue traversing the children when the parent was marked to be removed? To double assert that the previous loop worked as expected? (if that is the case, what about decoupling the previous loop into a separate function and adding unit test coverage instead of all the runtime `Assume` calls?).

Reply:
I see how this might be a performance improvement, but I don’t think this is going to be heavy enough to warrant this level of rewrite at this stage in the PR. Perhaps this could be done in a follow-up.